File: account-server-golang/server_api/MATCHMAKE.go
package server_api

import (
	"account-server/database"
	"account-server/gameservers"
	"account-server/resource_manager"
	"account-server/utils"
	"encoding/json"
	"errors"
	"time"
)

type request_MATCHMAKE struct {
	Request_id                string
	Session_id                string
	Game_name                 string
	Preferred_region          string
	Preferred_lobby_id        string
	Preferred_room_tag        string
	Game_client_resources_md5 string
}

// MATCHMAKE:
// 1. Busca un slot disponible en un gameserver y devuelve el ticket al cliente.
func MATCHMAKE(message []byte, sendResponse func([]byte)) error {

	var r request_MATCHMAKE

	opt_preferred_room_tag := "" // Parametro opcional (con valor por defecto "") para request_MATCHMAKE

	// Comprueba que estan todos los parametros requeridos.
	{
		var m map[string]any
		err := json.Unmarshal(message, &m)
		if err != nil {
			return err
		}
		if !utils.ContainsAllKeys(m, []string{"request_id", "session_id", "game_name", "preferred_region", "preferred_lobby_id", "game_client_resources_md5"}) {
			err = errors.New("MATCHMAKE bad input parameters")
			return err
		}

		if _, ok := m["require_test_instance"]; ok { // Evita que las requests, que indican require_test_instance, se ejecuten en una instancia del account server que no es de test.
			if !database.IsTestInstance() {
				return errors.New("require_test_instance")
			}
		}

		if v, ok := m["preferred_room_tag"]; ok {
			opt_preferred_room_tag = v.(string)
		}
	}

	err := json.Unmarshal(message, &r)
	if err != nil {
		return err
	}
	r.Preferred_room_tag = opt_preferred_room_tag

	var account dict

	// Comprueba si existe una cuenta con esa session_id
	{
		target := []string{"session_id", r.Session_id}
		responseDict, err := database.ReadAccount(target, "session_expire guest_account_id")
		if err != nil {
			return err
		}
		account = responseDict
	}

	responseDict := dict{}

	if account != nil {

		sessionExpired := false

		if account["guest_account_id"] != nil {
			// Para las cuentas guest no se aplica session_expire
		} else {
			// Comprueba que la session no esta caducada
			expireTimeStr := account["session_expire"].(string) // Sera "" si session_expire no esta en account
			if expireTimeStr == "" {
				panic(0)
			}
			expireTime, err := time.Parse(time.RFC3339, expireTimeStr)
			if err != nil {
				panic(0)
			}
			sessionExpired = expireTime.Before(time.Now())
		}

		if !sessionExpired {

			ticket, lobbyId := gameservers.MatchMake(r.Game_name, r.Preferred_region, r.Preferred_lobby_id, r.Preferred_room_tag, r.Session_id)

			// Si hay ticket entonces lo envia al cliente.
			if ticket != "" {
				if lobbyId == "" {
					panic(0)
				}

				gss := gameservers.GetGameServer(lobbyId)
				if gss != nil {
					responseDict["ticket"] = ticket

					game_map_names := make([]string, 0, 32)
					for _, k := range resource_manager.GetGameEnabledMaps(r.Game_name) {
						game_map_names = append(game_map_names, k.(string))
					}

					game_client_resources_md5 := resource_manager.GetGameClientResourcesMd5(r.Game_name)

					gs_info := dict{
						"lobby_id":                  gss.Lobby_id,
						"build_id":                  gss.Build_id,
						"ipv4_address":              gss.Ipv4_address,
						"address":                   gss.Address,
						"port":                      gss.Port,
						"region":                    gss.Region,
						"player_count":              gss.PlayerCount,
						"max_rooms":                 gss.Max_rooms,
						"max_players_per_room":      gss.Max_players_per_room,
						"game_name":                 gss.GameName,
						"game_client_resources_md5": game_client_resources_md5,
						"game_map_names":            game_map_names, // Nombres de todos los mapas del juego que estan listados en la entrada "ENABLED_MAPS" en el fichero game.cfg.
					}

					if game_client_resources_md5 != r.Game_client_resources_md5 {
						gs_info["game_client_resources"] = resource_manager.GetGameClientResources(r.Game_name)
					}

					responseDict["session_id"] = r.Session_id
					responseDict["gameserver_info"] = gs_info
					responseDict["request_id"] = r.Request_id
				} else {
					responseDict["error"] = "Gameserver no longer exists."
				}
			} else {
				responseDict["error"] = "All game servers are busy."
			}
		} else {
			responseDict["error"] = "Session is expired."
		}
	} else {
		responseDict["error"] = "No account found with this session_id."
	}

	// Envia la respuesta

	responseDict["request_id"] = r.Request_id

	response, err := json.Marshal(responseDict)
	if err != nil {
		return err
	}
	sendResponse(response)

	return nil
}

File: account-server/server_api/MATCHMAKE.py
import json
import datetime
from utils import generate_random_str

# MATCHMAKE:
# 1. Busca un slot disponible en un gameserver y devuelve el ticket al cliente.
async def MATCHMAKE(req, ws, send_message, accounts_cache, gameservers_info, resource_manager):

    # Aqui se debe comprobar que la req es correcta y en caso contrario se ignora sin enviar nada al cliente.
    if not all(k in req for k in ('request_id', 'session_id', 'game_name', 'game_client_resources_md5')):
        print("ERROR: MATCHMAKE bad input parameters.")
        return

    error_response = {
        "error": "Invalid ticket.",
        "request_id": req['request_id']
    }    

    session_id = req['session_id']

    ### Comprueba que exite el session_id
    query = {"session_id": session_id}
    projection = 'session_id'
    response = await accounts_cache.read_account(req, ws, send_message, query, projection)
    if response is None:
        return

    valid = False

    if 'session_id' in response:
    
        gsinstances = gameservers_info['instances']

        ### Comprueba que no exista ya un ticket para ese session_id
        ticket = None
        exit_loop = False
        for lobby_id, gs in gsinstances.items():
            if exit_loop:
                break
            for tk, tk_data in gs['tickets'].items():
                if tk_data['session_id'] == session_id:
                    ticket = tk
                    exit_loop = True
                    break

            for tk, ss_id in gs['players'].items():
                if ss_id == session_id:
                    ticket = tk
                    exit_loop = True
                    break
        
        ### Si no existe ya un ticket entonces crea uno nuevo.
        if not ticket:
            # Busca los game servers que tienen algun slot libre y que estan sirviendo
            # el juego 'game_name'.
            gss = [(lobby_id, gs) for lobby_id, gs in gsinstances.items() 
                if len(gs['players']) < gs['max_player_count'] and
                gs['game']['game_name'] == req['game_name']]

            if gss:
                # Elige el primero de la lista
                selected_index = 0
                lobby_id = gss[selected_index][0]
                gs = gss[selected_index][1]

                tickets = gs['tickets']

                ticket = generate_random_str(32)
                tickets[ticket] = {
                    'session_id': session_id,
                    'expire_time': datetime.datetime.now() + gameservers_info['ticket_timeout']
                }
                print('Tickets asignados {}/{}'.format(len(gs['players']), gs['max_player_count']))
                print('Tickets pendientes de validacion {}'.format(len(tickets)))

        ### Si hay ticket entonces lo envia al cliente.
        if ticket:
            response['ticket'] = ticket

            game_client_resources_md5 = resource_manager.get_game_client_resources_md5(req['game_name'])
            
            # game_resources_md5 = gs['game']['resources_md5']

            gameserver_info = {
                'lobby_id': lobby_id,
                'build_id': gs['build_id'],
                'ipv4_address': gs['ipv4_address'],
                'port': gs['port'],
                'player_count': len(gs['players']),
                'max_player_count': gs['max_player_count'],
                'game_name': gs['game']['game_name'],
                'game_client_resources_md5': game_client_resources_md5,
                'game_map_names': [x for x in resource_manager.get_game_maps(req['game_name']).keys()] # Nombres de todos los mapas del juego
            }

            if game_client_resources_md5 != req['game_client_resources_md5']:
                gameserver_info['game_client_resources'] = resource_manager.get_game_client_resources(req['game_name'])

            response['gameserver_info'] = gameserver_info

            response["request_id"] = req['request_id']
            await send_message(response, ws)   # send response

            valid = True
        else:
            error_response['error'] = "All game servers are busy."
    else:
        error_response['error'] = "No account found with this session_id."

    if not valid:
        await send_message(error_response, ws)   # send response

    